File: framework/headless/carousel.go
package headless

import (
	"strconv"
	"strings"

	"github.com/DonaldMurillo/gofastr/core-ui/html"
	"github.com/DonaldMurillo/gofastr/core/render"
)

// The carousel: a labelled region whose track is a real scrollable
// list of slides. Without script the track scrolls (native
// scroll-snap) and the controls are fragment anchors to the slide
// ids; with script the module owns the active slide (aria-current +
// a class), the prev/next stepping, the dot anchors, the slide status
// through the server's sentence, the auto-rotation (paused on hover,
// focus, hidden tab and reduced motion), and the reduced-motion gate.

// Carousel parts.
const (
	PartCarouselStage Part = "carousel-stage"
	PartCarouselTrack Part = "carousel-track"
	PartCarouselSlide Part = "carousel-slide"
	PartCarouselDot   Part = "carousel-dot"
	PartCarouselPrev  Part = "carousel-prev"
	PartCarouselNext  Part = "carousel-next"
)

// CarouselSlide is one entry.
type CarouselSlide struct {
	// Content is the slide body. Required.
	Content render.HTML
	// Label is the slide's accessible label. Defaults to
	// Strings.CarouselSlide ("Slide <n> of <total>") at render.
	Label string
	// ID is the slide's stable id. Empty takes "<trackID>-slide-<n>".
	ID string
}

// CarouselProps configures the carousel.
type CarouselProps struct {
	// Label names the region. Required.
	Label string
	// Slides, in order. Required, at least one.
	Slides []CarouselSlide
	// Loop makes Next on the last wrap to the first (and vice versa).
	// Default: the ends disable.
	Loop bool
	// AutoRotateMS, when > 0, advances every N ms. Negative refused.
	AutoRotateMS int
	// VisiblePerView shows N slides side-by-side (default 1; 0 or
	// negative refused — a viewport of nothing is not a viewport).
	VisiblePerView int
	// NoDots hides the dot anchors.
	NoDots bool
	// NoArrows hides the prev/next controls.
	NoArrows bool

	ID         string
	ExtraAttrs html.Attrs

	// Parts: attrs on the root, the track, the slides and the
	// controls.
	Parts   Parts
	Strings *Strings
}

// Carousel renders the slider.
func Carousel(p CarouselProps, s Classes) render.HTML {
	checkLabel("Carousel", "Label", p.Label)
	if len(p.Slides) == 0 {
		panic("headless: Carousel requires at least one slide — an empty track is not a carousel")
	}
	if p.AutoRotateMS < 0 {
		panic("headless: Carousel AutoRotateMS " + strconv.Itoa(p.AutoRotateMS) + " is negative — a rotation ahead of the clock is not a rotation")
	}
	if p.VisiblePerView < 0 {
		panic("headless: Carousel VisiblePerView " + strconv.Itoa(p.VisiblePerView) + " is negative — a viewport narrower than nothing is not a viewport")
	}
	if p.VisiblePerView == 0 {
		p.VisiblePerView = 1
	}
	if p.VisiblePerView > 8 {
		// Carried data clamps, never refuses (the generated CSS covers
		// eight columns): a wider count renders the ceiling.
		p.VisiblePerView = 8
	}
	w := p.Strings.Resolve()

	b := p.Parts.Box(s)
	id := p.ID
	if id == "" {
		id = "hui-carousel-" + menuShortHash(p.Label)
	}
	trackID := id + "-track"

	// Slides: stable ids, labels through the Strings sentence. The
	// track and slides are DIVs, not ul/li: slides are not list items,
	// and a ul whose children carry role=group trips axe's list rule
	// besides the allowed-role one the coordinator's site sweep caught.
	ids := make([]string, 0, len(p.Slides))
	for i := range p.Slides {
		if p.Slides[i].ID == "" {
			p.Slides[i].ID = trackID + "-slide-" + strconv.Itoa(i)
		}
		checkFragmentID("Carousel slide", "ID", p.Slides[i].ID)
		ids = append(ids, p.Slides[i].ID)
	}
	checkNoDuplicateIDs("Carousel", ids)

	rootAttrs := Merge(Safe(p.ExtraAttrs, "role", "aria-roledescription"), html.Attrs{
		"role":                 "region",
		"aria-roledescription": "carousel",
		"aria-label":           scrubControlBytes(p.Label),
		"id":                   id,
	})
	Mark(rootAttrs, "data-hui-carousel")
	if p.AutoRotateMS > 0 {
		rootAttrs["data-hui-carousel-rotate-ms"] = strconv.Itoa(p.AutoRotateMS)
	}
	if p.Loop {
		Mark(rootAttrs, "data-hui-carousel-loop")
	}

	slides := make([]render.HTML, 0, len(p.Slides))
	for i, sl := range p.Slides {
		label := sl.Label
		if label == "" {
			label = carouselSlideName(w.CarouselSlide, i+1, len(p.Slides))
		}
		own := html.Attrs{
			"id":                   sl.ID,
			"role":                 "group",
			"aria-roledescription": "slide",
			"aria-label":           scrubControlBytes(label),
		}
		if i == 0 {
			own["aria-current"] = "true"
		}
		slides = append(slides, b.El("div", PartCarouselSlide, own, sl.Content))
	}

	trackAttrs := Attrs(map[string]string{
		"id":       trackID,
		"tabindex": "0",
	})
	Mark(trackAttrs, "data-hui-carousel-track")
	// The status format travels beside the initial sentence: the module
	// re-says the sentence through the format after every step, so the
	// words stay the server's while the number stays live.
	trackAttrs["data-hui-carousel-status"] = carouselSlideName(w.CarouselSlide, 1, len(p.Slides))
	trackAttrs["data-hui-carousel-status-fmt"] = w.CarouselSlide + "|" + strconv.Itoa(len(p.Slides))
	// The stage is the positioning context the overlaid controls
	// hang from: it wraps exactly the track, so an absolutely
	// positioned prev/next centres on the slides and can never grow
	// down over the dot row (a WCAG 2.2 target-size failure the flat
	// structure brought with it).
	stage := []render.HTML{b.El("div", PartCarouselTrack, trackAttrs, slides...)}
	if !p.NoArrows {
		last := len(p.Slides) - 1
		stage = append(stage,
			b.El("a", PartCarouselPrev, html.Attrs{
				"href":                   "#" + p.Slides[0].ID,
				"aria-label":             w.Previous,
				"data-hui-carousel-prev": "",
			}, render.Text(w.Previous)),
			b.El("a", PartCarouselNext, html.Attrs{
				"href":                   "#" + p.Slides[last].ID,
				"aria-label":             w.Next,
				"data-hui-carousel-next": "",
			}, render.Text(w.Next)),
		)
	}
	children := []render.HTML{b.El("div", PartCarouselStage, nil, stage...)}

	if !p.NoDots {
		dots := make([]render.HTML, 0, len(p.Slides))
		for i, sl := range p.Slides {
			own := html.Attrs{
				"href":                   "#" + sl.ID,
				"aria-label":             numReplace(w.CarouselGoTo, "{n}", i+1),
				"data-hui-carousel-goto": strconv.Itoa(i),
			}
			if i == 0 {
				own["aria-current"] = "true"
			}
			dots = append(dots, b.El("a", PartCarouselDot, own, render.Text(strconv.Itoa(i+1))))
		}
		children = append(children, b.El("nav", PartMarker, html.Attrs{
			"aria-label": scrubControlBytes(p.Label),
		}, dots...))
	}
	return b.El("div", PartRoot, rootAttrs, children...)
}

// carouselSlideName formats the slide sentence.
func carouselSlideName(format string, n, total int) string {
	return numReplace(numReplace(format, "{n}", n), "{total}", total)
}

// numReplace substitutes a {token} with a number.
func numReplace(s, token string, n int) string {
	return strings.ReplaceAll(s, token, strconv.Itoa(n))
}

File: framework/headless/carousel.js
// headless-carousel: the behaviour module for this package's
// carousels. The track is a native scrollable list (the no-script
// contract); the module owns the active slide — aria-current plus a
// class on exactly one slide, mirrored onto the dots — the prev/next
// stepping (looping when the loop flag rides the root), the
// keyboard (ArrowLeft/ArrowRight on the focused track, RTL-aware),
// the status sentence re-said through the server's words, and the
// auto-rotation with its pauses (hover, focus, hidden tab, reduced
// motion). It replaces the retired core-ui/runtime carousel module.
(function () {
  'use strict';
  const NAME = 'headless-carousel';
  const NS = window.__gofastr = window.__gofastr || {};
  if (NS.loadedModules && Object.prototype.hasOwnProperty.call(NS.loadedModules, NAME)) return;
  NS.loadedModules = NS.loadedModules || {};
  NS.loadedModules[NAME] = true;

  const REDUCED = window.matchMedia && window.matchMedia('(prefers-reduced-motion: reduce)');

  function within(root, sel) {
    const out = [];
    if (root.matches && root.matches(sel)) out.push(root);
    if (root.querySelectorAll) out.push.apply(out, root.querySelectorAll(sel));
    return out;
  }

  function trackOf(root) {
    // No :scope >: the stage wrapper sits between the root and the
    // track now, and the track stays the first descendant in document
    // order (a nested carousel's track is inside a slide of this one).
    return root.querySelector('[data-hui-carousel-track]');
  }
  function slidesOf(root) {
    return Array.from(root.querySelectorAll('[data-hui-carousel-track] > [aria-label]'));
  }
  function currentOf(root) {
    const slides = slidesOf(root);
    for (let i = 0; i < slides.length; i++) {
      if (slides[i].getAttribute('aria-current') === 'true') return i;
    }
    return 0;
  }

  // setActive marks exactly one slide and its dot, scrolls the track,
  // and re-says the status through the server's sentence shape.
  function setActive(root, idx) {
    const slides = slidesOf(root);
    if (slides.length === 0) return;
    idx = ((idx % slides.length) + slides.length) % slides.length;
    for (let i = 0; i < slides.length; i++) {
      const on = i === idx;
      if (on) slides[i].setAttribute('aria-current', 'true');
      else slides[i].removeAttribute('aria-current');
    }
    for (const dot of root.querySelectorAll('[data-hui-carousel-goto]')) {
      if (parseInt(dot.getAttribute('data-hui-carousel-goto'), 10) === idx) {
        dot.setAttribute('aria-current', 'true');
      } else {
        dot.removeAttribute('aria-current');
      }
    }
    const track = trackOf(root);
    if (track) {
      const sl = slides[idx];
      if (sl && track.scrollLeft !== sl.offsetLeft - track.offsetLeft) {
        try { sl.scrollIntoView({ block: 'nearest', inline: 'start', behavior: REDUCED && REDUCED.matches ? 'auto' : 'smooth' }); } catch (_) {}
      }
      // The status sentence: the format and total travel on the
      // -fmt hook ("<sentence>|<total>"); the words stay the server's.
      const pack = track.getAttribute('data-hui-carousel-status-fmt');
      if (pack) {
        const bar = pack.indexOf('|');
        if (bar > 0) {
          const fmt = pack.slice(0, bar);
          const total = pack.slice(bar + 1);
          track.setAttribute('data-hui-carousel-status',
            fmt.replace(/\{n\}/g, String(idx + 1)).replace(/\{total\}/g, total));
        }
      }
    }
  }

  // ─── auto-rotation ───────────────────────────────────────────────
  //
  // One timer per rotating carousel at its own cadence (4000 when the
  // rotate attribute carries none), stepping only when !paused(root).
  // Created on arm, cleared on disarm — which runs on gofastr:navigate
  // and from the detach watcher below. No shared sweep, no 1 Hz poll:
  // the retired module's shape (one interval per carousel, stopped on
  // hidden and reduced motion) is the model, with the hover/focus
  // pauses checked at each tick instead of by a listener pair.
  function cadenceOf(root) {
    const ms = parseInt(root.getAttribute('data-hui-carousel-rotate-ms'), 10);
    return Number.isFinite(ms) && ms > 0 ? ms : 4000;
  }

  function paused(root) {
    if (document.hidden) return true;
    if (REDUCED && REDUCED.matches) return true;
    if (root.matches(':hover') || root.contains(document.activeElement)) return true;
    return false;
  }

  // A Set, not a WeakSet: the detach sweep below has to walk every
  // rotating root to find the ones that left the tree, and a WeakSet
  // cannot be walked — which is how the watcher came to query the
  // DOCUMENT for connected roots and then test them for !isConnected,
  // a condition that is never true. Iterability is the fix, and it
  // costs nothing: the set holds at most one entry per live carousel.
  const rotating = new Set(); // roots with a live timer

  function stepOnce(root) {
    if (!root.isConnected || paused(root)) return;
    setActive(root, currentOf(root) + 1);
  }

  function armRotation(root) {
    if (!root.hasAttribute('data-hui-carousel-rotate-ms')) return;
    if (rotating.has(root)) return;
    rotating.add(root);
    root.__huiTimer = setInterval(function () { stepOnce(root); }, cadenceOf(root));
  }

  function disarmRotation(root) {
    if (!rotating.has(root)) return;
    rotating.delete(root);
    if (root.__huiTimer) { clearInterval(root.__huiTimer); root.__huiTimer = null; }
  }

  // A DETACHED carousel never fires another tick we can see, but its
  // interval keeps the element (and its slides) alive: watch for the
  // detach directly, the same childList-only watcher shape the
  // disclosure trap uses, and disarm what went away. The sweep walks
  // the tracked set, not the document: a root that left the tree is
  // by definition no longer in querySelectorAll's answer, which is
  // why the previous connected-query + !isConnected sweep never fired.
  let detachWatcher = null;
  const disarmDetached = () => {
    for (const root of Array.from(rotating)) {
      if (!root.isConnected) disarmRotation(root);
    }
  };
  const syncDetachWatcher = () => {
    if (rotating.size > 0 && !detachWatcher) {
      detachWatcher = new MutationObserver(disarmDetached);
      detachWatcher.observe(document.body, { childList: true, subtree: true });
    } else if (rotating.size === 0 && detachWatcher) {
      detachWatcher.disconnect();
      detachWatcher = null;
    }
  };

  // ─── controls ────────────────────────────────────────────────────

  function clampStep(root, idx) {
    const n = slidesOf(root).length;
    const loop = root.hasAttribute('data-hui-carousel-loop');
    if (idx < 0 || idx >= n) return loop ? ((idx % n) + n) % n : -1;
    return idx;
  }

  document.addEventListener('click', function (e) {
    const t = e.target;
    if (!t || !t.closest) return;
    const root = t.closest('[data-hui-carousel]');
    if (!root) return;
    const prev = t.closest('[data-hui-carousel-prev]');
    const next = t.closest('[data-hui-carousel-next]');
    const dot = t.closest('[data-hui-carousel-goto]');
    if (prev) {
      e.preventDefault();
      const to = clampStep(root, currentOf(root) - 1);
      if (to >= 0) setActive(root, to);
      return;
    }
    if (next) {
      e.preventDefault();
      const to = clampStep(root, currentOf(root) + 1);
      if (to >= 0) setActive(root, to);
      return;
    }
    if (dot) {
      e.preventDefault();
      setActive(root, parseInt(dot.getAttribute('data-hui-carousel-goto'), 10));
      return;
    }
  });

  // Keyboard on the focused track: the arrows step (RTL-aware),
  // Home/End jump.
  document.addEventListener('keydown', function (e) {
    const track = e.target && e.target.closest && e.target.closest('[data-hui-carousel-track]');
    if (!track) return;
    const root = track.closest('[data-hui-carousel]');
    if (!root) return;
    if (e.key !== 'ArrowLeft' && e.key !== 'ArrowRight' && e.key !== 'Home' && e.key !== 'End') return;
    const rtl = getComputedStyle(track).direction === 'rtl';
    const cur = currentOf(root);
    let to = -1;
    if (e.key === 'ArrowRight') to = clampStep(root, cur + ((rtl) ? -1 : 1));
    else if (e.key === 'ArrowLeft') to = clampStep(root, cur + ((rtl) ? 1 : -1));
    else if (e.key === 'Home') to = 0;
    else if (e.key === 'End') to = slidesOf(root).length - 1;
    if (to >= 0) {
      e.preventDefault();
      setActive(root, to);
    }
  });

  // ─── the arrival pass ────────────────────────────────────────────

  function scan(root) {
    const scope = root && root.querySelectorAll ? root : document;
    for (const el of within(scope, '[data-hui-carousel]')) {
      armRotation(el);
    }
    syncDetachWatcher();
  }

  window.addEventListener('gofastr:navigate', function () {
    for (const root of Array.from(document.querySelectorAll('[data-hui-carousel-rotate-ms]'))) {
      disarmRotation(root);
    }
    syncDetachWatcher();
  });

  scan(document);
  NS._moduleScanners = NS._moduleScanners || {};
  NS._moduleScanners[NAME] = scan;
})();
